File: algorithm-families/String Algorithms/Python/code.py
from collections import Counter

def string_length(string):
    length = 0

    for _ in string:
        length += 1

    return length

def reverse_string(string):
    characters = list(string)

    left = 0
    right = string_length(characters) - 1

    while left < right:
        characters[left], characters[right] = (
            characters[right],
            characters[left],
        )

        left += 1
        right -= 1

    return "".join(characters)

def is_palindrome(string):
    left = 0
    right = string_length(string) - 1

    while left < right:
        if string[left] != string[right]:
            return False

        left += 1
        right -= 1

    return True

def character_frequency(string):
    frequency = {}

    for character in string:
        if character in frequency:
            frequency[character] += 1
        else:
            frequency[character] = 1

    return frequency

def naive_pattern_matching(text, pattern):
    matches = []

    n = string_length(text)
    m = string_length(pattern)

    if m == 0 or m > n:
        return matches

    for i in range(n - m + 1):
        j = 0

        while j < m and text[i + j] == pattern[j]:
            j += 1

        if j == m:
            matches.append(i)

    return matches

def longest_common_prefix(strings):
    if not strings:
        return ""

    prefix = strings[0]

    for string in strings[1:]:
        length = 0

        while (
            length < string_length(prefix)
            and length < string_length(string)
            and prefix[length] == string[length]
        ):
            length += 1

        prefix = prefix[:length]

        if not prefix:
            break

    return prefix

def longest_common_subsequence(first, second):
    n = string_length(first)
    m = string_length(second)

    dp = [
        [0] * (m + 1)
        for _ in range(n + 1)
    ]

    for i in range(1, n + 1):
        for j in range(1, m + 1):
            if first[i - 1] == second[j - 1]:
                dp[i][j] = (
                    dp[i - 1][j - 1] + 1
                )
            else:
                dp[i][j] = max(
                    dp[i - 1][j],
                    dp[i][j - 1],
                )

    return dp[n][m]

def build_lps(pattern):
    m = string_length(pattern)

    lps = [0] * m

    length = 0
    i = 1

    while i < m:
        if pattern[i] == pattern[length]:
            length += 1
            lps[i] = length
            i += 1

        elif length != 0:
            length = lps[length - 1]

        else:
            lps[i] = 0
            i += 1

    return lps


def kmp_pattern_matching(text, pattern):
    matches = []

    n = string_length(text)
    m = string_length(pattern)

    if m == 0 or m > n:
        return matches

    lps = build_lps(pattern)

    i = 0
    j = 0

    while i < n:
        if text[i] == pattern[j]:
            i += 1
            j += 1

        if j == m:
            matches.append(i - j)

            j = lps[j - 1]

        elif i < n and text[i] != pattern[j]:
            if j != 0:
                j = lps[j - 1]
            else:
                i += 1

    return matches

def are_anagrams(first, second):
    if string_length(first) != string_length(second):
        return False

    frequency = {}

    for character in first:
        if character in frequency:
            frequency[character] += 1
        else:
            frequency[character] = 1

    for character in second:
        if character not in frequency:
            return False

        frequency[character] -= 1

        if frequency[character] < 0:
            return False

    return True

def compress_string(string):
    if not string:
        return ""

    compressed = []

    i = 0
    length = string_length(string)

    while i < length:
        current = string[i]
        count = 0

        while (
            i < length
            and string[i] == current
        ):
            count += 1
            i += 1

        compressed.append(current)
        compressed.append(str(count))

    return "".join(compressed)

def main():

    text = "Hello, World!"

    print("String:", text)
    print("Length:", string_length(text))

    print("----------------------------------------")

    reversed_text = "algorithm"

    print("Original:", reversed_text)

    reversed_text = reverse_string(reversed_text)

    print("Reversed:", reversed_text)

    print("----------------------------------------")

    palindrome = "racecar"

    print(
        f'Palindrome "{palindrome}":',
        is_palindrome(palindrome),
    )

    print("----------------------------------------")

    frequency_text = "hello world"

    print(
        f'Character Frequency for "{frequency_text}":'
    )

    frequency = character_frequency(
        frequency_text
    )

    for character, count in frequency.items():
        print(f"'{character}': {count}")

    print("----------------------------------------")

    search_text = "abracadabra"
    pattern = "abra"

    print("Naive Pattern Matching:")
    print("Text:", search_text)
    print("Pattern:", pattern)

    matches = naive_pattern_matching(
        search_text,
        pattern,
    )

    print("Indices:", *matches)

    print("----------------------------------------")

    words = [
        "flower",
        "flow",
        "flight",
    ]

    prefix = longest_common_prefix(words)

    print("Longest Common Prefix:", prefix)

    print("----------------------------------------")

    first = "ABCBDAB"
    second = "BDCABA"

    lcs = longest_common_subsequence(
        first,
        second,
    )

    print(
        "Longest Common Subsequence Length:",
        lcs,
    )

    print("----------------------------------------")

    kmp_text = "ABABDABACDABABCABAB"
    kmp_pattern = "ABABCABAB"

    print("KMP Pattern Matching:")
    print("Text:", kmp_text)
    print("Pattern:", kmp_pattern)

    matches = kmp_pattern_matching(
        kmp_text,
        kmp_pattern,
    )

    print("Indices:", *matches)

    print("----------------------------------------")

    anagram_a = "listen"
    anagram_b = "silent"

    print(
        f'Are "{anagram_a}" and "{anagram_b}" '
        f"anagrams?",
        are_anagrams(
            anagram_a,
            anagram_b,
        ),
    )

    print("----------------------------------------")

    compression_input = "aaabbcdddd"

    compressed = compress_string(
        compression_input
    )

    print("Original:", compression_input)
    print("Compressed:", compressed)

    print("----------------------------------------")


if __name__ == "__main__":
    main()

File: algorithm-families/String Algorithms/C++/code.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <limits>

using namespace std;

int string_length(const string& str)
{
    int length = 0;

    for (char character : str)
    {
        (void)character;
        length++;
    }

    return length;
}

void reverse_string(string& str)
{
    int left = 0;
    int right = string_length(str) - 1;

    while (left < right)
    {
        char temp = str[left];
        str[left] = str[right];
        str[right] = temp;

        left++;
        right--;
    }
}

bool is_palindrome(const string& str)
{
    int left = 0;
    int right = string_length(str) - 1;

    while (left < right)
    {
        if (str[left] != str[right])
        {
            return false;
        }

        left++;
        right--;
    }

    return true;
}

void character_frequency(const string& str)
{
    unordered_map<char, int> frequency;

    for (char character : str)
    {
        frequency[character]++;
    }

    for (const auto& [character, count] : frequency)
    {
        cout << "'" << character << "': "
             << count << '\n';
    }
}

vector<int> naive_pattern_matching(
    const string& text,
    const string& pattern
)
{
    vector<int> matches;

    int n = string_length(text);
    int m = string_length(pattern);

    if (m == 0 || m > n)
    {
        return matches;
    }

    for (int i = 0; i <= n - m; i++)
    {
        int j = 0;

        while (
            j < m &&
            text[i + j] == pattern[j]
        )
        {
            j++;
        }

        if (j == m)
        {
            matches.push_back(i);
        }
    }

    return matches;
}

string longest_common_prefix(
    const vector<string>& strings
)
{
    if (strings.empty())
    {
        return "";
    }

    string prefix = strings[0];

    for (size_t i = 1; i < strings.size(); i++)
    {
        int j = 0;

        while (
            j < static_cast<int>(prefix.length()) &&
            j < static_cast<int>(strings[i].length()) &&
            prefix[j] == strings[i][j]
        )
        {
            j++;
        }

        prefix = prefix.substr(0, j);

        if (prefix.empty())
        {
            break;
        }
    }

    return prefix;
}

int longest_common_subsequence(
    const string& first,
    const string& second
)
{
    int n = string_length(first);
    int m = string_length(second);

    vector<vector<int>> dp(
        n + 1,
        vector<int>(m + 1, 0)
    );

    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= m; j++)
        {
            if (first[i - 1] == second[j - 1])
            {
                dp[i][j] =
                    dp[i - 1][j - 1] + 1;
            }
            else
            {
                dp[i][j] =
                    max(
                        dp[i - 1][j],
                        dp[i][j - 1]
                    );
            }
        }
    }

    return dp[n][m];
}

vector<int> build_lps(const string& pattern)
{
    int m = string_length(pattern);

    vector<int> lps(m, 0);

    int length = 0;
    int i = 1;

    while (i < m)
    {
        if (pattern[i] == pattern[length])
        {
            length++;
            lps[i] = length;
            i++;
        }
        else if (length != 0)
        {
            length = lps[length - 1];
        }
        else
        {
            lps[i] = 0;
            i++;
        }
    }

    return lps;
}


vector<int> kmp_pattern_matching(
    const string& text,
    const string& pattern
)
{
    vector<int> matches;

    int n = string_length(text);
    int m = string_length(pattern);

    if (m == 0 || m > n)
    {
        return matches;
    }

    vector<int> lps =
        build_lps(pattern);

    int i = 0;
    int j = 0;

    while (i < n)
    {
        if (text[i] == pattern[j])
        {
            i++;
            j++;
        }

        if (j == m)
        {
            matches.push_back(i - j);

            j = lps[j - 1];
        }
        else if (
            i < n &&
            text[i] != pattern[j]
        )
        {
            if (j != 0)
            {
                j = lps[j - 1];
            }
            else
            {
                i++;
            }
        }
    }

    return matches;
}

bool are_anagrams(
    const string& first,
    const string& second
)
{
    if (first.length() != second.length())
    {
        return false;
    }

    unordered_map<char, int> frequency;

    for (char character : first)
    {
        frequency[character]++;
    }

    for (char character : second)
    {
        frequency[character]--;

        if (frequency[character] < 0)
        {
            return false;
        }
    }

    return true;
}

string compress_string(const string& str)
{
    if (str.empty())
    {
        return "";
    }

    string compressed;

    int i = 0;

    while (i < string_length(str))
    {
        char current = str[i];
        int count = 0;

        while (
            i < string_length(str) &&
            str[i] == current
        )
        {
            count++;
            i++;
        }

        compressed += current;
        compressed += to_string(count);
    }

    return compressed;
}

void print_separator()
{
    cout << "----------------------------------------\n";
}


void print_matches(const vector<int>& matches)
{
    for (int index : matches)
    {
        cout << index << ' ';
    }

    cout << '\n';
}

int main()
{

    string text = "Hello, World!";

    cout << "String: "
         << text
         << '\n';

    cout << "Length: "
         << string_length(text)
         << '\n';

    print_separator();

    string reversed = "algorithm";

    cout << "Original: "
         << reversed
         << '\n';

    reverse_string(reversed);

    cout << "Reversed: "
         << reversed
         << '\n';

    print_separator();

    string palindrome = "racecar";

    cout << "Palindrome \""
         << palindrome
         << "\": "
         << (
                is_palindrome(palindrome)
                    ? "true"
                    : "false"
            )
         << '\n';

    print_separator();

    string frequency_text =
        "hello world";

    cout << "Character Frequency for \""
         << frequency_text
         << "\":\n";

    character_frequency(
        frequency_text
    );

    print_separator();

    string search_text =
        "abracadabra";

    string pattern =
        "abra";

    cout << "Naive Pattern Matching:\n";

    cout << "Text: "
         << search_text
         << '\n';

    cout << "Pattern: "
         << pattern
         << '\n';

    cout << "Indices: ";

    print_matches(
        naive_pattern_matching(
            search_text,
            pattern
        )
    );

    print_separator();

    vector<string> words = {
        "flower",
        "flow",
        "flight"
    };

    cout << "Longest Common Prefix: "
         << longest_common_prefix(words)
         << '\n';

    print_separator();

    string first =
        "ABCBDAB";

    string second =
        "BDCABA";

    cout << "Longest Common Subsequence Length: "
         << longest_common_subsequence(
                first,
                second
            )
         << '\n';

    print_separator();

    string kmp_text =
        "ABABDABACDABABCABAB";

    string kmp_pattern =
        "ABABCABAB";

    cout << "KMP Pattern Matching:\n";

    cout << "Text: "
         << kmp_text
         << '\n';

    cout << "Pattern: "
         << kmp_pattern
         << '\n';

    cout << "Indices: ";

    print_matches(
        kmp_pattern_matching(
            kmp_text,
            kmp_pattern
        )
    );

    print_separator();

    string anagram_a =
        "listen";

    string anagram_b =
        "silent";

    cout << "Are \""
         << anagram_a
         << "\" and \""
         << anagram_b
         << "\" anagrams? "
         << (
                are_anagrams(
                    anagram_a,
                    anagram_b
                )
                    ? "true"
                    : "false"
            )
         << '\n';

    print_separator();

    string compression_input =
        "aaabbcdddd";

    string compressed =
        compress_string(
            compression_input
        );

    cout << "Original: "
         << compression_input
         << '\n';

    cout << "Compressed: "
         << compressed
         << '\n';

    print_separator();


    return 0;
}